### ISSUE TITLE
False positive with redundant_closure

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust-clippy#4354]
Not sure if it's not a duplicate of some issue, there are a lot of tickets, but this still seems distinct. Tested both on nightly and stable.

```
~/k/pastebinrun (master|✔) $ cargo clippy -V
clippy 0.0.212 (082cfa7 2019-06-07)
~/k/pastebinrun (master|✔) $ cargo +nightly clippy -V
clippy 0.0.212 (b041511 2019-08-07)
```

Sample code:

```rust
fn callbacker(_: impl Fn(&mut [u8])) {}
fn g<W>(_: W) {}
fn main() {
    callbacker(|w| g(w));
}
```

Causes the following warning.

```
    Checking playground v0.0.1 (/playground)
warning: redundant closure found
 --> src/main.rs:4:16
  |
4 |     callbacker(|w| g(w));
  |                ^^^^^^^^ help: remove closure as shown: `g`
  |
  = note: `#[warn(clippy::redundant_closure)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

The closure cannot be removed as shown.

**Comment by apiraino** [comment]
I think I've found another test case:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=74a6ad88e84f7403129a7cd96801b666
```
fn main() {
    // forcing cast to String for test purposese
    let _v : Vec<String> = match Some("test,test1,test2") {
        Some(x) => x
            .split(',')
            // clippy suggests this closure is redundant
            .map(|x| String::from(x))
            .collect(),
        None => vec![],
    };
}```

**Comment by xfix** [comment]
@apiraino This is correct, the following code works:

```rust
fn main() {
    // forcing cast to String for test purposese
    let _v : Vec<String> = match Some("test,test1,test2") {
        Some(x) => x
            .split(',')
            .map(String::from)
            .collect(),
        None => vec![],
    };
    
}
```

**Comment by apiraino** [comment]
oh I am sorry, I had completely misunderstood the suggestion:
```
6 |             .map(|x| String::from(x))
  |                  ^^^^^^^^^^^^^^^^^^^ help: remove closure as shown: `String::from`
```
